src/ga4gh/vrs/normalize.py

```diff
@@ -3,60 +3,192 @@
+) -> Optional[LocationPos]:
+    """Get Allele location start or end value for interval
+
+    :param allele_vo: VRS Allele object
+    :param use_start: `True` if using `allele_vo.location.start`. `False` if using
+        `allele_vo.location.end`
+    :return: A `LocationPos` if using integer or indefinite range. Otherwise return
+        `None`
     """
-    Converts .location.sequence into an IRI if it is a SequenceReference because it makes the code simpler.
-    If it started as a sequence reference, put it back as one at the end.
+    if use_start:
+        pos = allele_vo.location.start
+    else:
+        pos = allele_vo.location.end
+
+    if isinstance(pos, int):
+        val = pos
+        pos_type = PosType.INTEGER
+    else:
+        pos0_is_none = pos.root[0] is None
+        pos1_is_none = pos.root[1] is None
+
+        if not pos0_is_none and not pos1_is_none:  # definite range
+            return None
+
+        val = pos.root[0] or pos.root[1]
+        pos_type = PosType.RANGE_LT_OR_EQUAL if pos0_is_none else PosType.RANGE_GT_OR_EQUAL
+
+    return LocationPos(value=val, pos_type=pos_type)
+
+
+def _get_new_allele_location_pos(
+    new_pos_val: int, pos_type: PosType
+) -> Union[int, models.Range]:
+    """Get updated location pos on normalized allele
+
+    :param new_pos_val: New position after normalization
+    :param pos_type: Original position type used in pre-normalized VRS Allele object
+    :return: Updated position as integer or VRS Range object
+    """
+    if pos_type == PosType.INTEGER:
+        val = new_pos_val
+    elif pos_type == PosType.RANGE_LT_OR_EQUAL:
+        val = models.Range([None, new_pos_val])
+    else:
+        val = models.Range([new_pos_val, None])
+    return val
+
+
+def _normalize_allele(input_allele, data_proxy, rle_seq_limit=50):
+    """Normalize Allele using "fully-justified" normalization adapted from NCBI's
+    VOCA. Fully-justified normalization expands such ambiguous representation over the
+    entire region of ambiguity, resulting in an unambiguous representation that may be
+    readily compared with other alleles.
+
+    :param input_allele: Input VRS Allele object
+    :param data_proxy: SeqRepo dataproxy
+    :param rle_seq_limit: If RLE is set as the new state, set the limit for the length
+        of the `sequence`.
+        To exclude `sequence` from the response, set to 0.
+        For no limit, set to `None`.
+
+    Does not attempt to normalize Allele's with definite ranges. Will return the
+        `input_allele`
     """
     allele = pydantic_copy(input_allele)
+
+    # Temporarily convert SequenceReference to IRI because it makes the code simpler.
+    # This will be changed back to SequenceReference at the end of the method
     sequence_reference = None
     if isinstance(allele.location.sequence, models.SequenceReference):
         sequence_reference = allele.location.sequence
         allele.location.sequence = models.IRI(sequence_reference.refgetAccession)
```

Comment:
Not following this. An IRI isn't guaranteed to have the digest in it, so we should always assume a `SequenceReference`, or have a mechanism for resolving the IRI to get a `SequenceReference`. I might be missing something here, but why not simply remove this block and then revise line 107 to pull from the `digest` field expected in every object?

```suggestion
```

Reply:
This was initially done by @theferrit32 . @ahwagner is the digest expected in every VRS object? In the models.py, the `digest` is set as an optional field. 

I'm trying to figure out why this was done, but I'm now realizing that this is an issue if `allele.location.sequence` is not defined. A `SequenceLocation.sequence` is listed as an optional field. @ahwagner can you explain why this is not a required field?

Reply:
>@ahwagner is the digest expected in every VRS object?
Yes, this is something that I think @theferrit32 was working on. It is available to every object, and our digest strategy in VRS 2 will compute this for every object, whether or not it is an identifiable object. I believe we were going to be using this for all objects in VRS-Python.

>A SequenceLocation.sequence is listed as an optional field. @ahwagner can you explain why this is not a required field?
Yes, it is optional to use this attribute in JSON Schema, because when used in an `Allele` that is part of a `Haplotype`, the `SequenceLocation.sequence` can be omitted from VRS messages, as they (by definition) will match the `sequence` of the parent `Haplotype` object. However, it is required from the `ga4ghDigest.keys` for creating computed digests, because it is still a critical component of the value of a SequenceLocation. In those cases, it is expected that the system loading the VRS `Haplotype` object would refer to / copy over the `Haplotype.sequence` for the `Haplotype.member[*].location.sequence` properties.

Reply:
@ahwagner I think there might be some confusion regarding the `digest` field. At the moment, `digest` is optional and does not get computed when a VRS Object is created. Should this field be populated each time a user creates a VRS Object?

Reply:
@ahwagner is suggesting to switch this condition to IRI. Some test cases: `#seqrefs/myseq123` and `HTTPS://w3id.org/NM_012345`

Reply:
We will need to dereference the `#seqrefs/myseq123` outside this function because this function only receives the allele to be normalized, it won't have access to the full original document the allele came from, where `#seqrefs/myseq123` can be resolved from. 

We can create another function like `ga4gh_inline` (or something) that takes a JSON document, finds the GA4GH objects in it, and for any field whose value is an IRI that is a relative pointer, inline the object it points to in that field, if the document contains it.

Reply:
If we do that the `type` fields would then need to be required on the input JSON/dict. They are not currently required because the type fields are defined as literals and when you construct a particular class with some input it assumes it is that type and fills in the type field.

If we don't want this constraint, we could just traverse the input document and replace *all* field values that look like JSON pointers (not just those in fields that are defined as IRIs in the VRS models) with the objects they refer to. This would also let people use the JSON pointer thing in non-model fields. Like if someone has a statement and a custom field they added that isn't in the model, but they want to refer to the variant in the same document from there using a pointer.

Reply:
@theferrit32 updated comments + tests with deferenced IRIs. Let me know if I need to make any more changes!